controller/state.go

```diff
@@ -801,6 +800,51 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *v1
 	return &compRes, nil
 }
 
+// useDiffCache will determine if the diff should be calculated based
+// on the existing live state cache or not.
+func useDiffCache(noCache bool, manifestInfos []*apiclient.ManifestResponse, sources []v1alpha1.ApplicationSource, app *v1alpha1.Application, manifestRevisions []string, statusRefreshTimeout time.Duration) bool {
+
+	if noCache {
+		log.Debug("useDiffCache(false): noCache is true")
+		return false
+	}
+	_, refreshRequested := app.IsRefreshRequested()
+	if refreshRequested {
+		log.Debug("useDiffCache(false): refreshRequested")
+		return false
+	}
+	if app.Status.Expired(statusRefreshTimeout) {
+		log.Debug("useDiffCache(false): app.status.expired")
+		return false
+	}
+
+	revisionChanged := len(manifestInfos) != len(sources) ||
+		!reflect.DeepEqual(app.Status.GetRevisions(), manifestRevisions)
+	if revisionChanged {
+		log.Debug("useDiffCache(false): revisionChanged")
+		return false
+	}
+
+	currentSource := v1alpha1.ApplicationSource{}
+	if app.Spec.Source != nil {
+		currentSource = *app.Spec.Source
+	}
+	currentSpec := v1alpha1.ComparedTo{
+		Source:            currentSource,
+		Destination:       app.Spec.Destination,
+		Sources:           app.Spec.Sources,
+		IgnoreDifferences: app.Spec.IgnoreDifferences,
+	}
```

Comment:
It looks like each place that populates the `comparedTo` field currently toggles on `hasMultipleSources`. So the behavior here could be subtly different if, for example, someone foolishly specifies both the `source` and `sources` field.

How would you feel about centralizing `v1alpha1.ComparedTo` construction in a new `NewComparedTo` func? That should avoid any subtle differences in how the field is constructed.

Reply:
done